Generate Java call graphs from local JARs in `OPALPlugin`

mir-am opened the pull request:
## Description
This PR makes the following changes:
- Changes the `DataWriter` interface to `DataRW` so that the plugins can read/write from/to the disk given a base directory. 
- `OPALPlugin` first tries to generate a call graph from a local JAR file if available. If not, it downloads the JAR file from Maven central like before.

## Motivation and context
`POMAnalyzer` stores JAR files to the disk when resolving dependencies. Therefore, it is desirable to generate CGs from local JAR files and also reduce the network activity of `OPALPlugin`. 

## Testing
Tested with the DC.

proksch commented on core/src/main/java/eu/fasten/core/data/opal/MavenCoordinate.java:
```diff
@@ -156,9 +156,15 @@ public String toProductUrl() {
     }
+
+    public String toPath() {
+        return Paths.get(groupID.replace('.', '/'), artifactID, versionConstraint,
```
This requires a unix file system. Please use the system specific separator constant.

proksch replied:
Stupid question, why not simply return a `Path`?

mir-am replied:
Because, when joining paths using `Paths.get()`, `String` is needed. I've changed the method to return `Path` now.

mir-am replied:
Addressed in [3b12da6](https://github.com/fasten-project/fasten/pull/459/commits/3b12da65de924db0fa46c81047276b27866a104a)

proksch commented on core/src/main/java/eu/fasten/core/data/opal/MavenCoordinate.java:
```diff
@@ -156,9 +156,15 @@ public String toProductUrl() {
         var repo = mavenRepos.get(0);
         return this.toURL(repo) + "/" + artifactID + "-" + versionConstraint + "." + packaging;
```
This has not been changed, but does not make sense. Which URL is this supposed to create?

mir-am replied:
The method `toProductURL()` is not written by me. Based on its usage, I can tell it is used to download Maven artifacts like JARs.

proksch commented on analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/OPALPlugin.java:
```diff
@@ -76,9 +79,18 @@ public void consume(String kafkaRecord, ProcessingLane l) {
-                        CHA, date, artifactRepository, ONLY_STATIC_CALLSITES);
-                long duration = currentTimeMillis() - startTime; 
+                try {
+                    this.graph = OPALPartialCallGraphConstructor.createPCGFromLocalJar(mavenCoordinate, date,
```
Do not use exception handling for programming logic! Concrete proposal: There is no need to have two separate methods, one `createPartialCG` method should be enough, which just takes a path to the jar to be analyzed. The only difference is where the path comes from: either `mavenCoord.getPath()` exists or the .jar is downloaded and a path to the downloaded file is provided. As the jar should *always* exist locally in our setup, I would be fine with just throwing an error when the file is missing.

proksch commented on analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/OPALPlugin.java:
```diff
@@ -76,9 +79,18 @@ public void consume(String kafkaRecord, ProcessingLane l) {
+                            Paths.get(baseDir, mavenCoordinate.toPath()).toString(),
+                            CHA, ONLY_STATIC_CALLSITES);
+                } catch (FileNotFoundException e) {
+                    logger.warn(e.getMessage());
```
When logging, make sure that you preserve the stack trace.

As elaborated before, we can just remove this catch handling for now...

proksch commented on analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/OPALPartialCallGraphConstructor.java:
```diff
@@ -127,12 +126,39 @@ public static PartialJavaCallGraph createPartialJavaCG(
+     * @return RevisionCallGraph of the given coordinate.
+     * @throws FileNotFoundException
+     */
+    public static PartialJavaCallGraph createPCGFromLocalJar(final MavenCoordinate coordinate,
```
Please merge the two `createPCG` methods into one. There is NO need for a case distinction, both should do exactly the same...

proksch reviewed (changes requested):
I think that the solution is sub-optimal... a case distinction that should be taken on a high level is dragged down through the various levels of the API. I have proposed a different strategy in the individual comments.

mir-am commented:
@proksch, thanks for the code review and comments.
I have addressed your comments. This is not the best solution but it leverages our existing JARs in `.m2` and reduces the network traffic. Also, it downloads JARs, if not present in FS.
